zkevm-circuits/src/taiko_pi_circuit_.rs

```diff
@@ -0,0 +1,832 @@
+            max_transactions_per_block: block.protocol_instance.max_transactions_per_block,
+            max_bytes_per_tx_list: block.protocol_instance.max_bytes_per_tx_list,
+            field9,
+            field10,
+            block_context: block.context.clone(),
+            chain_id: block.context.chain_id,
+        }
+    }
+
+    fn get_pi(&self) -> H256 {
+        let rpi_bytes = self.rpi_bytes();
+        let rpi_keccak = keccak256(rpi_bytes);
+        H256(rpi_keccak)
+    }
+
+    fn get_pi_hi_low(&self) -> [(&'static str, Option<Word>, Vec<u8>); 2] {
+        let pi_bytes = self.get_pi().to_fixed_bytes();
+        [
+            ("high_16_bytes_of_keccak_rpi", None, pi_bytes[..16].to_vec()),
+            ("low_16_bytes_of_keccak_rpi", None, pi_bytes[16..].to_vec()),
+        ]
+    }   
+}
+/// Config for PiCircuit
+#[derive(Clone, Debug)]
+pub struct TaikoPiCircuitConfig<F: Field> {
+    q_enable: Selector,
+    public_input: Column<Instance>, // equality
+    block_acc: Column<Advice>,
+    field_gadget: FieldBytesGadget<F>,
+
+    block_number: [Cell<F>;2], // Phase1
+    keccak_output:[Cell<F>;2], // Phase2, equality
+
+    block_table: BlockTable,
+    keccak_table: KeccakTable,
+    byte_table: ByteTable,
+
+    // To annotate columns at assignment for debug purpose 
+    col_configs: Vec<CellColumn<F, PiCellType>>,
+}
+
+/// Circuit configuration arguments
+pub struct TaikoPiCircuitConfigArgs<F: Field> {
+    /// BlockTable
+    pub block_table: BlockTable,
+    /// KeccakTable
+    pub keccak_table: KeccakTable,
+    /// ByteTable
+    pub byte_table: ByteTable,
+    /// Challenges
+    pub challenges: Challenges<Expression<F>>,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum PiCellType {
+    Storage1,
+    Storage2,
+    Byte,
+    Lookup(Table)
+}
+impl CellType for PiCellType {
+    fn byte_type() -> Option<Self> {
+        Some(Self::Byte)
+    }
+    fn storage_for_phase(phase: u8) -> Self {
+        match phase {
+            1 => PiCellType::Storage1,
+            2 => PiCellType::Storage2,
+            _ => unimplemented!()
+        }
+    }
+}
+impl Default for PiCellType {
+    fn default() -> Self {
+        Self::Storage1
+    }
+}
+
+
+impl<F: Field> SubCircuitConfig<F> for TaikoPiCircuitConfig<F> {
+    type ConfigArgs = TaikoPiCircuitConfigArgs<F>;
+
+        /// Return a new TaikoPiCircuitConfig
+        fn new(
+            meta: &mut ConstraintSystem<F>,
+            Self::ConfigArgs {
+                block_table,
+                keccak_table,
+                byte_table,
+                challenges,
+            }: Self::ConfigArgs,
+        ) -> Self {
+            let field_cm = CellManager::new(
+                meta,
+                vec![
+                    (PiCellType::Byte, 32, 1, false),
+                    (PiCellType::Storage1, 3, 1, false)
+                ],
+                0,
+                1,
```

Comment:
I think with a single CellManager it will be easier to just make this height be whatever, and then the other code should be able to use the same height in `rot` to check between the different states, and in the witness assignment.